src/Acuminator/Acuminator.Analyzers/Refactorings/EventHandlerSignature/EventHandlerSignatureRefactoring.cs

```diff
@@ -0,0 +1,160 @@
+
+namespace Acuminator.Analyzers.Refactorings.EventHandlerSignature
+{
+	[ExportCodeRefactoringProvider(LanguageNames.CSharp), Shared]
+	public class EventHandlerSignatureRefactoring : CodeRefactoringProvider
+	{
+		private const string ArgsParameterName = "e";
+		private const string EventHandlerMethodName = "_";
+
+		public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+		{
+			context.CancellationToken.ThrowIfCancellationRequested();
+
+			var semanticModel = await context.Document.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
+			var pxContext = new PXContext(semanticModel.Compilation);
+
+			var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+			var methodNode = root?.FindNode(context.Span)?.GetDeclaringMethodNode();
+
+			if (methodNode != null)
+			{
+				IMethodSymbol methodSymbol = semanticModel.GetDeclaredSymbol(methodNode, context.CancellationToken);
+
+				if (IsSuitableForRefactoring(methodSymbol, pxContext))
+				{
+					var eventHandlerInfo = methodSymbol.GetEventHandlerInfo(pxContext);
+
+					if (eventHandlerInfo.eventSignatureType == EventHandlerSignatureType.Default
+						&& pxContext.Events.EventHandlerSignatureTypeMap.TryGetValue(
+							(eventHandlerInfo.eventType, EventHandlerSignatureType.Generic), out var genericArgsSymbol)
+						&& methodSymbol.Name.EndsWith("_" + eventHandlerInfo.eventType, StringComparison.Ordinal))
+					{
+						(string dacName, string fieldName) = ParseMethodName(methodSymbol.Name, eventHandlerInfo.eventType);
+
+						string title = nameof (Resources.EventHandlerSignatureCodeActionTitle).GetLocalized().ToString();
+						context.RegisterRefactoring(CodeAction.Create(title,
+							ct => ChangeSignatureAsync(context, root, semanticModel, methodNode, methodSymbol,
+								eventHandlerInfo.eventType, genericArgsSymbol, dacName, fieldName, ct), title));
+					}
+				}
+			}
+		}
+
+		private bool IsSuitableForRefactoring(IMethodSymbol symbol, PXContext pxContext)
+		{
+			return symbol?.ContainingType?.OriginalDefinition != null
+			       && symbol.ContainingType.OriginalDefinition.IsPXGraphOrExtension(pxContext)
+			       && symbol.Parameters.Length <= 2
+			       && !symbol.IsOverride
+				   && !symbol.GetAttributes().Any(
+					   attr => attr.AttributeClass != null && attr.AttributeClass.Equals(pxContext.AttributeTypes.PXOverrideAttribute));
+		}
+
+		private Task<Document> ChangeSignatureAsync(CodeRefactoringContext context, 
+			SyntaxNode root, SemanticModel semanticModel,
+			MethodDeclarationSyntax methodDeclaration, IMethodSymbol methodSymbol,
+			EventType eventType, INamedTypeSymbol genericArgsSymbol,
+			string dacName, string fieldName,
+			CancellationToken cancellationToken)
+		{
+			cancellationToken.ThrowIfCancellationRequested();
+
+			var newParameters = methodDeclaration.ParameterList.Parameters;
+			var cacheParameterSymbol = methodSymbol.Parameters[0];
```

Comment:
@dropsonic, the `IsSuitableForRefactoring` method checks that the number of parameters in method <= 2
Are there some other checks which filter out methods with zero parameters? If not then such method will cause an exception.

Reply:
@SENya1990 it is checked later using `GetEventHandlerInfo`